collectors/mimecast/utils.js

```diff
@@ -28,39 +33,53 @@ function getAPILogs(authDetails, state, accumulator, maxPagesPerInvocation) {
                     if (error) {
                         return reject(error);
                     }
                     if (response.statusCode &&
-                         (response.statusCode === 429 ||
-                          response.statusCode >= 500 )) {
+                        (response.statusCode === 429 ||
+                            response.statusCode >= 500)) {
                         return reject(response);
                     }
-                    try {
-                        body = JSON.parse(body);
-                    } catch (exception) {
-                        AlLogger.error("MIME000010 Error parsing response. ", body);
-                        return reject(exception);
-                    }
-                    if (body.fail && body.fail[0] && body.fail[0].errors) {
-                        return reject(body.fail[0].errors[0]);
+
+                    if (!compressFlag) {
+                        try {
+                            body = JSON.parse(body);
+                        } catch (exception) {
+                            AlLogger.error("MIME000010 Error parsing response. ", body);
+                            return reject(exception);
+                        }
+                        if (body.fail && body.fail[0] && body.fail[0].errors) {
+                            return reject(body.fail[0].errors[0]);
+                        }
                     }
+
                     pageCount++;
                     switch (state.stream) {
                         case Siem_Logs:
-                            if (body.data && body.data.length > 0) {
-                                accumulator.push(...body.data);
-                            }
-                            AlLogger.debug(`MIME000011 accumulated first element: ${JSON.stringify(accumulator[1])}`);
+                            //get zip file store it in /tmp/mimecast/siemlog
+                            //unzip in same folder
+                            //for each file in /tmp/mimecast/siemlog get json data into accumator push
+                            fs.mkdir(SIEM_LOGS_PATH, { recursive: true }, (err) => {
+                                if (err) throw err;
+                                unzipBuffer(body).then(function (bodyData) {
+                                    accumulator.push(...bodyData);
+                                    AlLogger.debug(`MIME000011 accumulated first element: ${JSON.stringify(accumulator[1])} and accumaulator length ${accumulator.length}`);
+                                    removeTempDirectory(SIEM_LOGS_PATH);
+                                }).catch((error) => {
+                                    AlLogger.debug(`MIME000011 Error Accumulating Data: ${error} compress flagged: ${compressFlag}`);
+                                });
+                            });
                             if (response.meta && response.meta.isLastToken) {
```

Comment:
we need to move this `if` statement into `unzipBuffer()` callback function. Otherwise `resolve({ accumulator, nextPage })` can happen before we actually call `accumulator.push(...bodyData)`, line 73